test/mip_test.c

```diff
@@ -49,9 +67,108 @@ static void test_statechange(void) {
   onstatechange(&iface);
 }
 
+// *****************************************************************************
+// Implement tap_rx() tap_tx() tap_up()
+
+static size_t tap_rx(void *buf, size_t len, void *userdata) {
+  ssize_t received = read((int) userdata, buf, len); // Original
+  usleep(1);  // This is to avoid 100% CPU
+  if (received < 0) return 0;
+  return (size_t) received;
+}
+
+static size_t tap_tx(const void *buf, size_t len, void *userdata) {
+  ssize_t res = write((int) userdata, buf, len); // Original
+  if (res < 0) {
+    MG_ERROR(("tap_tx failed: %d", errno));
+    return 0;
+  }
+  return (size_t) res;
+}
+
+static bool tap_up(void *userdata) {
+  return userdata ? true : false;
+}
+
+// *****************************************************************************
+// HTTP fetch
+
+static void f_http_fetch(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
+  int *fn_data_int_ptr = (int *) fn_data;
+  // MG_INFO(("%d", ev));
+  if (ev == MG_EV_CONNECT) {
+    // c->is_hexdumping = 1;
+    mg_printf(c, "GET /%s HTTP/1.0\r\nHost: %s\r\n\r\n",
+              c->rem.is_ip6 ? "" : "/robots.txt",
+              c->rem.is_ip6 ? "ipv6.google.com" : "cesanta.com");
+  } else if (ev == MG_EV_HTTP_MSG) {
+    struct mg_http_message *hm = (struct mg_http_message *) ev_data;
+    // MG_INFO(("-->[%.*s]", (int) hm->message.len, hm->message.ptr));
+    // ASSERT(mg_vcmp(&hm->method, "HTTP/1.1") == 0);
+    // ASSERT(mg_vcmp(&hm->uri, "301") == 0);
+    *fn_data_int_ptr = mg_http_status(hm);
+  } else if (ev == MG_EV_CLOSE) {
+    if (*fn_data_int_ptr == 0) *fn_data_int_ptr = 888;
+  } else if (ev == MG_EV_ERROR) {
+    if (*fn_data_int_ptr == 0) *fn_data_int_ptr = 777;
```

Comment:
That's a copy/paste from the unit_test which is not used in this test. Please remove EVERYTHING that is not used. Leave only relevant code.